### ISSUE TITLE
Concurrency exception when trying to insert entity with datetime2(0) key

### ISSUE THREAD

**Issue by clement911** [dotnet/efcore#27513]
Here is bug in EF6 Core.

Create an entity as follows:
```
public class QueryRun
    {
        [Column(TypeName = "datetime2(0)")]
        public DateTime StartedAt { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [NonDefault]
        public int Id { get; private set; }

        //other properties omitted for brevity
     }
```

Then try insert an entity
```
ctx.Add(new QueryRun
                {
                    StartedAt = DateTime.UtcNow,
                    //other properties omitted for brevity
                });
 await ctx.SaveChangesAsync();
```

This fails with the message: Expected 1 affected rows but got 0 rows!
After scratching my head for some time, I found the bug in the generated SQL code.

![image](https://user-images.githubusercontent.com/3426504/155903876-263d2ffd-42e2-4fe4-8249-5f78a36c0816.png)

We can see that  EF generated a param @p0 datetime2(7) instead of datetime2(0).
The issue is that the predicate `[StartedAt] = @p0` returns FALSE because the parameter has milliseconds whereas the column is datetime2(0) and has no milliseconds.

Expected behaviour: when generating the p0 parameter, EF should use the same type as the column. That is, it should use datetime2(0) instead of datetime2(7). If I change the generated code to use datetime2(0), I can see that the generated ID is returned.

**Comment by roji** [comment]
@clement911 can you confirm that you've confirmed StartAt as a concurrency token? Assuming so, what technique are you using to update it? In any case, a fully runnable code sample (including the model) is always important for us to investigate, rather than isolated snippets.

**Comment by clement911** [comment]
@roji on, StartsAt is not a concurrency token.

I've created a fully runnable repro for you here: https://github.com/clement911/EFBug

**Comment by roji** [comment]
Duplicate of https://github.com/dotnet/SqlClient/issues/1380

**Comment by roji** [comment]
This is because of a SqlClient issue specifically with scale 0 on datetime2, see https://github.com/dotnet/SqlClient/issues/1380. This issue is specifically called out in their [docs](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlparameter.scale?view=dotnet-plat-ext-5.0#remarks):

> For the DateTime2 type, scale 0 (the default) will be passed as datetime2(7). There is currently no way to send a parameter as datetime2(0). Scales 1-7 work as expected. This problem applies to DateTimeOffset and Time as well.

So while EF Core properly sets the SqlParameter's scale to 0, that gets ignored by SqlClient, which sends the full, untruncated value. Since that doesn't match the database column value (which is truncated), a concurrency exception occurs.

**Comment by roji** [comment]
In case it's useful for the future, here's an EF Core regression test for this:

<details>
<summary>Regression test</summary>

```c#
[ConditionalFact] // #27513
public void Insert_with_concurrent_token_with_non_default_store_type()
{
    using var testStore = SqlServerTestStore.CreateInitialized(DatabaseName);
    using (var context = new BlogContextWithConcurrencyTokenWithNonDefaultStoreType(testStore.Name))
    {
        context.Database.EnsureCreatedResiliently();

        var blog = new Blog
        {
            CreatedOn = new DateTime(2000, 1, 1, 1, 1, 1, millisecond: 1),
            Name = "original"
        };
        context.Blogs.Add(blog);
        context.SaveChanges();

        blog.Name = "modified";
        context.SaveChanges();
    }
}

public class BlogContextWithConcurrencyTokenWithNonDefaultStoreType : ContextBase
{
    public BlogContextWithConcurrencyTokenWithNonDefaultStoreType(string databaseName)
        : base(databaseName)
    {
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        base.OnModelCreating(modelBuilder);

        modelBuilder
            .Entity<Blog>()
            .Property(e => e.CreatedOn)
            .HasColumnType("datetime2(0)")
            .IsConcurrencyToken();
    }
}
```

</details>

**Comment by clement911** [comment]
@roji , understood about the SqlClient bug.
However in my case there isn't even a concurrency token column but there is an key identity column, which causes EF to query the DB to fetch the generated ID.
So, this issue is not constrained to scenarios with a concurrency token.
Any time you have a composite key with a datetime2(0) column and a identify column, the issue will occur upon insertion.

**Comment by clement911** [comment]
FYI, our work around for now is to truncate the datetime value in the client c# code. That is, we remove the milliseconds part BEFORE calling savechanges.

**Comment by clement911** [comment]
I suppose EF could apply that work around internally.
I.e. if the column is of type datetime2(0), it could set the milliseconds part to 0 before saving. I'm not sure if that could cause other problems though...

**Comment by roji** [comment]
> However in my case there isn't even a concurrency token column but there is an key identity column, which causes EF to query the DB to fetch the generated ID.

Yes, that's right. The way inserts are currently implemented when there's an IDENTITY column, EF first sends the INSERT, and then a SELECT to get back generated column values for the newly-inserted row (i.e. the identity value); this is why the SqlClient bug manifests even when there's no concurrency token. Note that #27573 changes how inserts work for 7.0, so that this should no longer affect inserts (except when triggers are defined on the table), but the bug would still manifest if there's a concurrency token.

> FYI, our work around for now is to truncate the datetime value in the client c# code. That is, we remove the milliseconds part BEFORE calling savechanges.

Yes, that's the workaround for now. I'd even recommend implementing the truncation in the .NET property setter, e.g.:

```c#
        [Column(TypeName = "datetime2(0)")]
        public DateTime StartedAt
        {
            get => _startedAt;
            set => /* truncate and set */ 
        }

        private DateTime _startedAt;
```

> I suppose EF could apply that work around internally.

We generally try to avoid compensating for bugs in lower levels, though we'll discuss this specific case in triage.

**Comment by roji** [comment]
Reopening as the bug is not going to be fixed on the SqlClient side (https://github.com/dotnet/SqlClient/issues/1380#issuecomment-1331183446). If we implement client-side parameter transformations (#28028), we can perform truncations in EF to work around this.

**Comment by ajcvickers** [comment]
Note from triage: we won't be doing anything specifically in EF for this. Workarounds are to manually truncate the type before calling SaveChanges, to truncate in overridden SaveChanges method or in SavingChanges event, or use an interceptor.

**Comment by clement911** [comment]
Looks like SqlClient now has a fix in the latest preview release:
https://github.com/dotnet/SqlClient/pull/2411

I hope future EF can leverage it to fix this bug.

**Comment by clement911** [comment]
Will this not be implemented?

**Comment by roji** [comment]
@AndriySvyryd any context on why we closed this?

**Comment by AndriySvyryd** [comment]
The underlying issue was fixed in SqlClient 6.0.1 and since 7.0 EF doesn't even issue SQL that is affected by this.

**Comment by clement911** [comment]
@AndriySvyryd I don't quite get what you're saying.
Was the issue fixed in EF7?

**Comment by AndriySvyryd** [comment]
@clement911 Yes, at least the one specifically shown in your repro.

**Comment by clement911** [comment]
@AndriySvyryd awesome.
I just tested again and can confirm that this works in EF9 👍

**Comment by EamonHetherton** [comment]
I just wanted to add some context here as I implemented the "fix" in SqlClient, and I'm not sure it is as "fixed" in EF as this comment chain implies, at least not as a result of the change in SqlClient. (EF9 also still has a dependency on Microsoft.Data.SqlClient >= 5.1.6 so wouldn't even pick up the zero scale fix for https://github.com/dotnet/SqlClient/pull/2411 without a user explicitly referencing Microsoft.Data.SqlClient >= 6.01 )

I originally came across the bug with 0 scale in SqlClient via EF which is why I think it is very relevant here.  My migration would create tables with datetime2(0) columns but then query them with datetime2(7) parameters.  I was also using partitioning on these tables and due to the way SQL server would only do partition elimination if the parameter type exactly matched the partition function type, I would get massive additional IO when the parameter type was datetime2(7) instead of datetime2(0)

I thought the change was uncontroversial; if a user specifies scale=0, the SqlClient should emit a parameter with scale=0.  However, I had a lot of pushback on changing the default behaviour in SqlClient to correctly emit 0 scale when the user explicitly set it.  Eventually I relented and implemented the fix behind a context switch to move it forward so that I could at least get access to the correct functionality even if SqlClient continued to use the original behaviour by default.  The documentation ([https://learn.microsoft.com/en-us/sql/connect/ado-net/appcontext-switches?view=sql-server-ver16](https://learn.microsoft.com/en-us/sql/connect/ado-net/appcontext-switches) has not yet updated to include the new context switch but I have left some suggested content [here](https://github.com/dotnet/SqlClient/pull/2411#issuecomment-2401105698)

So in summary, the problem still exists even when using EF9 unless the following two things are also true:
1) you are referencing Microsoft.Data.SqlClient >= 6.01
2) you set the context switch like this: 
`AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.LegacyVarTimeZeroScaleBehaviour", false);`

I've tested this with EF9 and Microsoft.Data.SqlClient v6.01, it will still emit datetime2(7) parameters even when the property has had its "Precision" set to 0 unless the context switch above is set to false. (why we use Precision in EF to set Scale in SqlClient is another matter entirely)

Also, if anyone carries some weight with the SqlClient team and can push for the fix to become the default behaviour in SqlClient that would be great :)
